src/v/storage/compaction_reducers.cc

```diff
@@ -181,8 +170,21 @@ copy_data_segment_reducer::filter(model::record_batch batch) {
     // 1. compute which records to keep
     std::vector<int32_t> offset_deltas;
     offset_deltas.reserve(batch.record_count());
+    int last_batch_records_seen = 0;
     co_await batch.for_each_record_async(
-      [this, &batch, &offset_deltas](const model::record& r) {
+      [this, &batch, &offset_deltas, &last_batch_records_seen](
+        const model::record& r) {
+          if (
+            batch.last_offset() == _segment_last_offset
+            && batch.record_count() == ++last_batch_records_seen) {
+              vlog(
+                stlog.trace,
+                "retaining last record: {} of segment from batch: {}",
+                r,
+                batch.header());
+              offset_deltas.push_back(r.offset_delta());
+              return ss::make_ready_future<>();
```

Comment:
When I'm looking at the description in the commit message, it all makes sense how the problem occurred. But I'm wondering why the fix involves more/different comparisons. For instance, we want to keep the last batch, which would have the ending offset 1799 like we want. The decision to keep the last batch: why does that involve offsets at all--it feels like it is a structural decision. It's physically the last batch, regardless of any content that the batch contains / represents.

Reply:
It is mostly a structural limitation of this code: we base compaction off of readers, and readers themselves are typically unaware of whether something is the last batch until _after_ we've iterated past it.

Reply:
makes sense!